text/0779-first-class-component-templates.md

```diff
@@ -0,0 +1,1622 @@
+
+Given this design, we can also simplify the definition of named components (both forms will of course be legal):
+
+```ts
+export interface GreetingArgs {
+  name: string;
+}
+
+export const Greeting = <template[GreetingArgs]>
+  <p>Hello, {{@name}}!</p>
+</template>
+```
+
+There are two key restrictions here:
+
+1. As mentioned above, this is illegal in the context of a class-backed component, because the component class itself is the host for the signature (and must be to make `this.args` type check correctly).
+
+2. The only thing allowed within the `[...]` is a type available in the local scope. *It is not legal to provide an inline type definition.* However, given the relative verbosity of even today’s component signature, still less the revised version from [RFC #0748][rfc-0748], inline signatures are unlikely to be attractive anyway.
+
+From an implementation perspective, this requires our language parser to handle this variant of the tag, and for the transforms supplied for compilation to properly ignore this for build output but to supply it in an appropriate place for TypeScript-aware tools like Glint to be able to take advantage of it.
+
+
+### Custom file extension
+
+These tooling considerations together provide the motivation for a custom file extension (`.gjs` and `.gts`). In the case of TypeScript in particular, it is not possible to *remove* errors using a TypeScript language server plugin, which means that in a pure `.js` or `.ts` file, a user would get conflicting reports from TypeScript and (e.g.) Glint. Thus, today, Glint recommends that GlimmerX users *disable TypeScript in their projects*, and rely on only Glint. Taking a lesson from Vue and Svelte, however, introducing a custom file extension allows us to provide a default type for `.gjs`/`.gts` files which makes TypeScript happy in `.js` and `.ts` files, and on top of which tools like Glint can safely add *more* information.[^hbs-custom-syntax]
+
+While both Prettier and ESLint *can* work with `.js` or `.ts`, introducing the new file extension also simplifies the tooling implementation for them. It does mean that tools like GitHub’s Linguist will not work without implementing support, but we need to do that work anyway.
+
+[^hbs-custom-syntax]: Note that this also applies to the `hbs` syntax discussed in [**Alternatives: Template literals (`hbs`)**](#template-literals-hbs).
+
+
+## How we teach this
+
+We describe a `<template>` as representing the *template* for a *component*. When there is no backing class, that’s all there is to the component. When there *is* a backing class, the component also has associated state and behavior. (Notably, this shift already began with Octane, where we generate template-only components by default.)
```

Comment:
> When there is no backing class, that’s all there is to the component

I think it might be useful to mention that this is "just a template only component, which we have taught for a long time"

Reply:
This is a great (better!) way of saying it in the context of addressing *existing* Ember users. Here, I intentionally chose phrasing oriented around someone new to Ember, and *not* in terms of existing terminology, in part because I think we should move increasingly to treating *class-backed* components as the more unusual case and “template-only” components as just “components”, the default. I have made this explicit as of ed5e12a.